bluetooth: Fix various plm issues

sean-madigan opened the pull request:
- callback_list was renamed to conn_cbs in commit 3eb975d. However plm callback was missed due to it not being built anywhere.
- Path loss monitoring was not being enabled anywhere, meaning that issues could be merged in without catching the issue. This commit adds a test case to build the shell with path loss monitoring enabled to catch issues.
- Path loss monitoring does not work without LE Power Control also enabled in the controller, so update the dependencies in the kconfigs to reflect this.

sean-madigan committed:
bluetooth: host: conn: Fix renamed callback_list

callback_list was renamed to conn_cbs in commit 3eb975d.

However plm callback was missed due to it not being built
anywhere.

Signed-off-by: Sean Madigan <sean.madigan@nordicsemi.no>

sean-madigan committed:
bluetooth: controller: kconfig: select power control for plm

Path loss monitoring does not work without LE Power Control
also enabled in the controller, so update the dependencies
in the kconfigs so path loss monitoring selects power
control.

Signed-off-by: Sean Madigan <sean.madigan@nordicsemi.no>

cvinayak commented on subsys/bluetooth/controller/Kconfig:
```diff
@@ -560,6 +560,7 @@ config BT_CTLR_LE_POWER_CONTROL
 	bool "LE Power Control Request Feature"
 	depends on BT_CTLR_LE_POWER_CONTROL_SUPPORT
 	default y if BT_TRANSMIT_POWER_CONTROL
+	default y if BT_PATH_LOSS_MONITORING
```
You should instead have `depends on BT_CTLR_LE_PATH_LOSS_MONITORING_SUPPORT && BT_CTLR_LE_POWER_CONTROL`  in `config BT_CTLR_LE_PATH_LOSS_MONITORING`, right?

sean-madigan replied:
Yeah that is probably good to have as well. Worth having this `default y if BT_PATH_LOSS_MONITORING` here though, otherwise it just silently doesn't work.

Thalley replied:
> otherwise it just silently doesn't work.

The Kconfig `depends on` should give build errors or warnings

sean-madigan replied:
Yeah that's what I thought but without this change if I just enable  `CONFIG_BT_PATH_LOSS_MONITORING=y` then I get no warning about the missing dependency for `BT_CTLR_LE_PATH_LOSS_MONITORING`. However, if I just try and enable `CONFIG_BT_CTLR_LE_PATH_LOSS_MONITORING=y` directly, then I do get the warning.

sean-madigan replied:
I presume this is for cases where `CONFIG_BT_CTLR=n` to not gives loads of warnings?

Thalley replied:
> Yeah that's what I thought but without this change if I just enable `CONFIG_BT_PATH_LOSS_MONITORING=y` then I get no warning about the missing dependency for `BT_CTLR_LE_PATH_LOSS_MONITORING`. However, if I just try and enable `CONFIG_BT_CTLR_LE_PATH_LOSS_MONITORING=y` directly, then I do get the warning.

Is that the case with or without `default y if BT_PATH_LOSS_MONITORING`? If you have `default y if BT_PATH_LOSS_MONITORING` then you probably don't see any warnings, but it should show one without. Did you try a clean build?

sean-madigan replied:
This is without `default y if BT_PATH_LOSS_MONITORING`, yes. And just double checked with a clean build

Thalley replied:
Sounds odd. Did it get enabled in the .config file?

sean-madigan replied:
No, not enabled in the .config

Thalley replied:
Sounds strange that if it doesn't get enabled but it set to `=y` in your config file and that you don't get any warnings. How are you building?

sean-madigan replied:
With these changes cherry picked to NCS, and `default y if BT_PATH_LOSS_MONITORING` removed, build the zephyr shell with `west build -b nrf52840dk/nrf52840 -p -- -DCONFIG_BT_PATH_LOSS_MONITORING=y` and can see `CONFIG_BT_PATH_LOSS_MONITORING=y` in the .config file but not `CONFIG_BT_CTLR_LE_PATH_LOSS_MONITORING=y`. With no warning printed.

If I do `west build -b nrf52840dk/nrf52840 -p -- -DCONFIG_BT_PATH_LOSS_MONITORING=y -DCONFIG_BT_CTLR_LE_PATH_LOSS_MONITORING=y` then I get the warning.

Thalley replied:
Weird. There really should be a warning if you request a Kconfig value that isn't applied. Sounds like a bug somewhere, and shouldn't be fixed by this approach IMO

sean-madigan replied:
Looks like we use `select` elsewhere in the file, so have updated to do this

cvinayak commented on tests/bluetooth/shell/testcase.yaml:
```diff
@@ -29,6 +29,13 @@ tests:
     build_only: true
+  bluetooth.shell.path_loss_monitoring:
+    extra_configs:
+      - CONFIG_BT_PATH_LOSS_MONITORING=y
```
```suggestion
      - CONFIG_BT_TRANSMIT_POWER_CONTROL=y
      - CONFIG_BT_PATH_LOSS_MONITORING=y
```

cvinayak replied:
Did you forget this?

sean-madigan replied:
What is the reason for this? Do you mean I should combine with the testcase above?

cvinayak replied:
i thought we were against `select BT_CTLR_LE_POWER_CONTROL` in favour of `depends on`.

Refer to: https://github.com/zephyrproject-rtos/zephyr/pull/70251#discussion_r1601487145

> In summary, here are some recommended practices for select:
Avoid selecting symbols with prompts or dependencies. Prefer depends on. If depends on causes annoying bloat in configuration files, consider adding a Kconfig default for the most common value.

@tejlmand FYI

sean-madigan replied:
As discussed [here](https://github.com/zephyrproject-rtos/zephyr/pull/76175#discussion_r1687276841), `depends on` does not have the desired effect - I presume this is why `select` has been used on other Kconfigs in the file

cvinayak replied:
> As discussed [here](https://github.com/zephyrproject-rtos/zephyr/pull/76175#discussion_r1687276841), `depends on` does not have the desired effect - I presume this is why `select` has been used on other Kconfigs in the file

"desired effect" has not been a criteria in the "recommended practices", adding new Kconfig default is another recommended practice (does add more Kconfig... but...) and hence will like @tejlmand to provide his comments.

Thalley commented on subsys/bluetooth/controller/Kconfig:
```diff
@@ -560,13 +560,14 @@ config BT_CTLR_LE_POWER_CONTROL
 config BT_CTLR_LE_PATH_LOSS_MONITORING
 	bool "LE Path Loss Monitoring Feature"
-	depends on BT_CTLR_LE_PATH_LOSS_MONITORING_SUPPORT
+	depends on BT_CTLR_LE_PATH_LOSS_MONITORING_SUPPORT && BT_CTLR_LE_POWER_CONTROL
```
What if `CONFIG_BT_CTLR=n`?

sean-madigan replied:
If `CONFIG_BT_CTLR=n` then `BT_CTLR_LE_PATH_LOSS_MONITORING_SUPPORT` shouldn't be enabled

sean-madigan committed:
tests: bluetooth: shell: Add test case to build shell with plm

Path loss monitoring was not being enabled anywhere, meaning
that issues could be merged in without catching the issue.

This commit adds a test case to build the shell with path
loss monitoring enabled to catch issues.

Signed-off-by: Sean Madigan <sean.madigan@nordicsemi.no>

sean-madigan commented:
Where are we up to with this PR @cvinayak? Does it require any changes do you think?

Thalley commented:
> Where are we up to with this PR @cvinayak? Does it require any changes do you think?

Ideally it would be refactored to use `depends on` rather than `select`, but I'll leave it up to @cvinayak

cvinayak commented:
> Where are we up to with this PR @cvinayak? Does it require any changes do you think?

We have not got any comments from @tejlmand, but approving for now as I am ok with use of `select` for this Kconfig option.
